src/gazebo_mavlink_interface.cpp

```diff
@@ -363,22 +477,40 @@ void GazeboMavlinkInterface::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf
   else {
     gzerr << "Unkown protocol version! Using v" << protocol_version_ << "by default \n";
   }
+
 }
 
 // This gets called by the world update start event.
 void GazeboMavlinkInterface::OnUpdate(const common::UpdateInfo&  /*_info*/) {
+
+  if (previous_imu_seq_ > 0) {
+    imu_message_ready_future_.wait();
+    // Once we're in the rythm of waiting for the next sample, we don't
+    // expect any skips between ImuCallback and OnUpdate.
+    assert(previous_imu_seq_ + 1 == last_imu_message_.seq());
+  }
+
+  // Reset the future/promise. We need to do this no matter what, otherwise
+  // we rist that the promise is set twice in a row on startup.
```

Comment:
typo: `risk`
And the way I see this you have a race condition here: you reset the promise, but the imu callback could already have been called another time (or does gazebo prevent this in any way?).
I find it strange that this callback runs on a different thread from the IMU callback.

Reply:
Hm, right. So we have to add a mutex as well?

Reply:
I see this in some of the tests:
```
terminate called after throwing an instance of 'std::future_error'
  what():  std::future_error: Promise already satisfied
Aborted (core dumped)
```
So I might just have to re-implement this using a condition variable.

Reply:
Yes I think it's better suited for that use-case.